src/wayland/seat/mod.rs

```diff
@@ -333,77 +361,114 @@ impl<T: 'static> SeatState<T> {
-    //         inner.send_all_caps();
-    //     }
-    // }
+// Touch
+impl<T> Seat<T> {
+    /// Adds the touch capability to this seat
+    ///
+    /// You are provided a [`TouchHandle`], which allows you to send input events
+    /// to this pointer. This handle can be cloned.
+    ///
+    /// Calling this method on a seat that already has a touch capability
+    /// will overwrite it, and will be seen by the clients as if the
+    /// touchscreen was unplugged and a new one was plugged in.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # extern crate wayland_server;
+    /// #
+    /// # use smithay::wayland::seat::Seat;
+    /// #
+    /// # let mut display = wayland_server::Display::new();
+    /// # let (mut seat, seat_global) = Seat::new(
+    /// #     &mut display,
+    /// #     "seat-0".into(),
+    /// #     None
+    /// # );
+    /// let touch_handle = seat.add_touch();
+    /// ```
+    pub fn add_touch(&mut self, dh: &mut DisplayHandle<'_>) -> TouchHandle {
+        let mut inner = self.arc.inner.lock().unwrap();
+        let touch = TouchHandle::new();
+        if inner.touch.is_some() {
+            // If there's already a tocuh device, remove it notify the clients about the change.
+            inner.touch = None;
+            inner.send_all_caps(dh);
+        }
+        inner.touch = Some(touch.clone());
+        inner.send_all_caps(dh);
+        touch
+    }
 
-    /// Checks whether a given [`WlSeat`](wl_seat::WlSeat) is associated with this [`Seat`]
-    pub fn owns(&self, seat: &wl_seat::WlSeat) -> bool {
-        let inner = self.arc.inner.lock().unwrap();
-        inner.known_seats.iter().any(|s| s == seat)
+    /// Access the touch device of this seat, if any.
+    pub fn get_touch(&self) -> Option<TouchHandle> {
+        self.arc.inner.lock().unwrap().touch.clone()
+    }
+
+    /// Remove the touch capability from this seat
+    ///
+    /// Clients will be appropriately notified.
+    pub fn remove_touch(&mut self, dh: &mut DisplayHandle<'_>) {
+        let mut inner = self.arc.inner.lock().unwrap();
+        if inner.touch.is_some() {
+            inner.touch = None;
+            inner.send_all_caps(dh);
+        }
     }
 }
 
-impl<T> ::std::cmp::PartialEq for SeatState<T> {
+impl<T> ::std::cmp::PartialEq for Seat<T> {
     fn eq(&self, other: &Self) -> bool {
         Arc::ptr_eq(&self.arc, &other.arc)
     }
 }
 
+/// A Seat handle
+/// This struct gives you access to the control of the
+/// capabilities of the associated seat.
+///
+/// It is directly inserted in the wayland display by its [`new`](Seat::new) method.
+///
+/// This is an handle to the inner logic, it can be cloned.
+///
+/// See module-level documentation for details of use.
+#[derive(Debug)]
+pub struct SeatState<T> {
+    pd: PhantomData<T>,
+}
+
+impl<T> Default for SeatState<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T> Clone for SeatState<T> {
+    fn clone(&self) -> Self {
```

Comment:
Hmm this seems like the wrong thing to be able to clone.

Reply:
Right